pkg/streamingpromql/operators/functions/rate_increase.go

```diff
@@ -272,3 +278,74 @@ func rateSeriesValidator() RangeVectorSeriesValidationFunction {
 		lastCheckedMetricName = metricName
 	}
 }
+
+func delta(step *types.RangeVectorStepData, rangeSeconds float64, emitAnnotation types.EmitAnnotationFunc) (float64, bool, *histogram.FloatHistogram, error) {
+	fHead, fTail := step.Floats.UnsafePoints()
+	fCount := len(fHead) + len(fTail)
+
+	hHead, hTail := step.Histograms.UnsafePoints()
+	hCount := len(hHead) + len(hTail)
+
+	if fCount > 0 && hCount > 0 {
+		// We need either at least two histograms and no floats, or at least two floats and no histograms to calculate a delta.
+		// Otherwise, emit a warning and drop this sample.
+		emitAnnotation(annotations.NewMixedFloatsHistogramsWarning)
+		return 0, false, nil, nil
+	}
+
+	if fCount >= 2 {
+		val := floatDelta(fCount, fHead, fTail, step.RangeStart, step.RangeEnd, rangeSeconds)
+		return val, true, nil, nil
+	}
+
+	if hCount >= 2 {
+		val, err := histogramDelta(hCount, hHead, hTail, step.RangeStart, step.RangeEnd, rangeSeconds, emitAnnotation)
+		if err != nil {
+			err = NativeHistogramErrorToAnnotation(err, emitAnnotation)
+			return 0, false, nil, err
+		}
+		return 0, false, val, nil
+	}
+
+	return 0, false, nil, nil
+}
+
+func floatDelta(fCount int, fHead []promql.FPoint, fTail []promql.FPoint, rangeStart int64, rangeEnd int64, rangeSeconds float64) float64 {
+	firstPoint := fHead[0]
+
+	var lastPoint promql.FPoint
+	if len(fTail) > 0 {
+		lastPoint = fTail[len(fTail)-1]
+	} else {
+		lastPoint = fHead[len(fHead)-1]
+	}
+
+	delta := lastPoint.F - firstPoint.F
+	return calculateFloatRate(false, false, rangeStart, rangeEnd, rangeSeconds, firstPoint, lastPoint, delta, fCount)
+}
+
+func histogramDelta(hCount int, hHead []promql.HPoint, hTail []promql.HPoint, rangeStart int64, rangeEnd int64, rangeSeconds float64, emitAnnotation types.EmitAnnotationFunc) (*histogram.FloatHistogram, error) {
+	firstPoint := hHead[0]
+
+	var lastPoint promql.HPoint
+	if len(hTail) > 0 {
+		lastPoint = hTail[len(hTail)-1]
+	} else {
+		lastPoint = hHead[len(hHead)-1]
+	}
+
+	if firstPoint.H.UsesCustomBuckets() != lastPoint.H.UsesCustomBuckets() {
+		return nil, histogram.ErrHistogramsIncompatibleSchema
```

Comment:
Could you please add a test for this case? You should be able to extend or borrow from `TestRateIncreaseAnnotations` in `engine_test.go` for this.

Reply:
Added in b9f89fb9dd2f586bc2a148cb3a04da2afce430b5

Reply:
Note that sprintf removed in 350a0e100af8f0fa2d5e9a45cbda160973a5a897